src/mds/MDSContext.h

```diff
@@ -135,11 +137,39 @@ class C_IO_Wrapper : public MDSIOContext
   C_IO_Wrapper(MDSRank *mds_, MDSInternalContextBase *wrapped_) : MDSIOContext(mds_), wrapped(wrapped_) {
     assert(wrapped != NULL);
   }
+
+  ~C_IO_Wrapper() {
+    delete wrapped;
+    wrapped = nullptr;
+  }
+
   virtual void finish(int r) {
     wrapped->complete(r);
```

Comment:
You need to set wrapped to null after calling complete().  Contexts either delete themselves in complete(), or we shouldn't be deleting them at all.